Accelerate multianewArray on P for 2D arrays

luke-li-2003 opened the pull request:
Supports the inlining of multianewArray on P when a new 2D array is created with the second dimension having a length of zero.

luke-li-2003 commented:
Related issue: https://github.com/eclipse-openj9/openj9/issues/2424

luke-li-2003 commented:
@zl-wang FYI

luke-li-2003 commented:
The code is working and as far as I can tell without error for arr[0][0], arr[n][0], and arr[n][n] (where it should correctly engage the out of line code).

luke-li-2003 commented:
Performance: where the load is basically me declaring a bunch of 2D arrays:

|length|default|accelerated|
|-|-|-|
|0|2.837|233.8|
|1|1.974|134.1|
|10|0.517|28.93|
|128|0.05|1.35|

The throughput unit is 1e6.

luke-li-2003 commented:
The offheap implementation is not done yet, but shouldn't be hard.

zl-wang commented:
i supposed the **"length"** above is the 1st dimension size (i.e.  **n**). the 2nd dimension size is also **n**? or it varied in your tests?

luke-li-2003 commented:
The 2nd dimension is zero as outlined in the problem statement. When the 2nd dimension is not zero, the performance of default and accelerated code are pretty much the same (since the accelerated code just call anyway).

luke-li-2003 commented:
In what is the weirdest bug I have seen, the evaluator wouldn't work if I set `-Xnoaot` or `-Xshareclasses:none`. All other options (like `-Xshareclasses:reset`) have no effect on it.

luke-li-2003 commented:
A `sanity.functional` test revealed why failure, which should be caused by https://github.com/eclipse-openj9/openj9/issues/21844 and not related to my code.

I will rebase the code and try again.

https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline-Build-Test-Personal/28072/console

luke-li-2003 commented:
Generated machine code, for reference:

```
    0x7fff8ecec238 00000040 [    0x7fff8c21a2a0]                   2    Label L0018:    ; (Start of internal control flow)
    0x7fff8ecec238 00000040 [    0x7fff8c21c5e0] 80020004          2    lwz     gr0, [gr2, 4]
    0x7fff8ecec23c 00000044 [    0x7fff8c21c730] 80e20000          2    lwz     gr7, [gr2, 0]
    0x7fff8ecec240 00000048 [    0x7fff8c21c7d0] 2c070000          2    cmpwi   cr0, gr7, 0
    0x7fff8ecec244 0000004c [    0x7fff8c21c870] 408200b4          2    bne     cr0, Label L0023
    0x7fff8ecec248 00000050 [    0x7fff8c21c910] 2c000000          2    cmpwi   cr0, gr0, 0
    0x7fff8ecec24c 00000054 [    0x7fff8c21c9b0] 40820030          2    bne     cr0, Label L0019
    0x7fff8ecec250 00000058 [    0x7fff8c21cb00] e86f0060          2    ld      gr3, [gr15, 96]
    0x7fff8ecec254 0000005c [    0x7fff8c21cba0] 38a30010          2    addi    gr5, gr3, 16
    0x7fff8ecec258 00000060 [    0x7fff8c21ccf0] e8cf0068          2    ld      gr6, [gr15, 104]
    0x7fff8ecec25c 00000064 [    0x7fff8c21cd90] 7c253000          2    cmpd    cr0, gr5, gr6
    0x7fff8ecec260 00000068 [    0x7fff8c21ce30] 41810098          2    bgt     cr0, Label L0023
    0x7fff8ecec264 0000006c [    0x7fff8c21cf80] f8af0060          2    std     [gr15, 96], gr5
    0x7fff8ecec268 00000070 [    0x7fff8c21d0d0] 90830000          2    stw     [gr3, 0], gr4
    0x7fff8ecec26c 00000074 [    0x7fff8c21d170] 38a00000          2    li      gr5, 0000000000000000
    0x7fff8ecec270 00000078 [    0x7fff8c21d2c0] 90a30004          2    stw     [gr3, 4], gr5
    0x7fff8ecec274 0000007c [    0x7fff8c21d410] 90a30008          2    stw     [gr3, 8], gr5
    0x7fff8ecec278 00000080 [    0x7fff8c21d4b0] 48000084          2    b       Label L0021     
    0x7fff8ecec27c 00000084 [    0x7fff8c21d540]                   2    Label L0019:    
    0x7fff8ecec27c 00000084 [    0x7fff8c21d5d0] 3ca03fff          2    lis     gr5, 0000000000003FFF
    0x7fff8ecec280 00000088 [    0x7fff8c21d670] 60a5ffff          2    ori     gr5, gr5, 65535
    0x7fff8ecec284 0000008c [    0x7fff8c21d710] 7c002840          2    cmplw   cr0, gr0, gr5
    0x7fff8ecec288 00000090 [    0x7fff8c21d7b0] 41810070          2    bgt     cr0, Label L0023
    0x7fff8ecec28c 00000094 [    0x7fff8c21d850] 5405103a          2    rlwinm  gr5, gr0, 0000000000000002, 00000000FFFFFFFC
    0x7fff8ecec290 00000098 [    0x7fff8c21d900] 38a5000f          2    addi    gr5, gr5, 15
    0x7fff8ecec294 0000009c [    0x7fff8c21d9a0] 54a50038          2    rlwinm  gr5, gr5, 0000000000000000, FFFFFFFFFFFFFFF8
    0x7fff8ecec298 000000a0 [    0x7fff8c21da50] 1cc00010          2    mulli   gr6, gr0, 16
    0x7fff8ecec29c 000000a4 [    0x7fff8c21daf0] 7cc62a14          2    add     gr6, gr6, gr5
    0x7fff8ecec2a0 000000a8 [    0x7fff8c21dc40] e86f0060          2    ld      gr3, [gr15, 96]
    0x7fff8ecec2a4 000000ac [    0x7fff8c21dce0] 7cc61a14          2    add     gr6, gr6, gr3
    0x7fff8ecec2a8 000000b0 [    0x7fff8c21de30] e8ef0068          2    ld      gr7, [gr15, 104]
    0x7fff8ecec2ac 000000b4 [    0x7fff8c21ded0] 7c263800          2    cmpd    cr0, gr6, gr7
    0x7fff8ecec2b0 000000b8 [    0x7fff8c21df70] 41810048          2    bgt     cr0, Label L0023
    0x7fff8ecec2b4 000000bc [    0x7fff8c21e0c0] f8cf0060          2    std     [gr15, 96], gr6
    0x7fff8ecec2b8 000000c0 [    0x7fff8c21e210] 90830000          2    stw     [gr3, 0], gr4
    0x7fff8ecec2bc 000000c4 [    0x7fff8c21e360] 90030004          2    stw     [gr3, 4], gr0
    0x7fff8ecec2c0 000000c8 [    0x7fff8c21e4b0] e8840058          2    ld      gr4, [gr4, 88]
    0x7fff8ecec2c4 000000cc [    0x7fff8c21e550] 7cc51a14          2    add     gr6, gr5, gr3
    0x7fff8ecec2c8 000000d0 [    0x7fff8c21e5f0] 38a30008          2    addi    gr5, gr3, 8
    0x7fff8ecec2cc 000000d4 [    0x7fff8c21e690] 7c0903a6          2    mtctr   gr0
    0x7fff8ecec2d0 000000d8 [    0x7fff8c21e730] 38e00000          2    li      gr7, 0000000000000000
    0x7fff8ecec2d4 000000dc [    0x7fff8c21e7d0]                   2    Label L0020:
    0x7fff8ecec2d4 000000dc [    0x7fff8c21e910] 90860000          2    stw     [gr6, 0], gr4
    0x7fff8ecec2d8 000000e0 [    0x7fff8c21ea60] 90e60004          2    stw     [gr6, 4], gr7
    0x7fff8ecec2dc 000000e4 [    0x7fff8c21ebb0] 90e60008          2    stw     [gr6, 8], gr7
    0x7fff8ecec2e0 000000e8 [    0x7fff8c21ec50] 78c0e8c2          2    rldicl  gr0, gr6, 000000000000003D, 1FFFFFFFFFFFFFFF
    0x7fff8ecec2e4 000000ec [    0x7fff8c21edb0] 90050000          2    stw     [gr5, 0], gr0
    0x7fff8ecec2e8 000000f0 [    0x7fff8c21ee50] 38a50004          2    addi    gr5, gr5, 4
    0x7fff8ecec2ec 000000f4 [    0x7fff8c21eef0] 38c60010          2    addi    gr6, gr6, 16
    0x7fff8ecec2f0 000000f8 [    0x7fff8c21ef90] 4200ffe4          2    bdnz    cr0, Label L0020
    0x7fff8ecec2f4 000000fc [    0x7fff8c21f030] 48000008          2    b       Label L0021
    0x7fff8ecec2f8 00000100 [    0x7fff8c21f0c0]                   2    Label L0023:
    0x7fff8ecec2f8 00000100 [    0x7fff8c21f150] 4800001c          2    b       Outlined Label L0022
    0x7fff8ecec2fc 00000104 [    0x7fff8c21f340]                   2    Label L0021:    ; (End of internal control flow)
```

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
I am pretty certain these chunk of code does nothing, since if a child is shared by the callNode and the node (first if block), then the register must have already been added to the post condition in the code block above.

However, the x evaluator included it, so I am leaving it here for now. Feedback appreciated.

zl-wang replied:
if what you mentioned is true, these blocks of code can only cause bugs, instead of nothing. you cannot add a register to a dep condition repeatedly.

luke-li-2003 replied:
You are right. I forgot that while on x there is union to handle repeated registers gracefully, there isn't on p.

luke-li-2003 replied:
I added a check so at least the same register won't be inserted twice. I'm  still not sure if this is needed at all, though.

luke-li-2003 replied:
I just worry that there might be a case where this is required that I haven't thought about. After all the x evaluator should not just have a block of useless code.

luke-li-2003 commented:
I have eliminated all the initialisation to zero to the heap since per our discussion offline they should already by 0.

luke-li-2003 commented:
The functional test is still running; here's the  result for the performance test:

|a[n][0]|baseline|newest|before|
|-|-|-|-|
|0|2.8M|243M|240M|
|1|2M|139M|134M|
|8|631K|39M|39M|
|128|51K|1.4M|1.37M|
|1024|6.4K|31.9K|31K|

zl-wang commented:
yes, give it a try.  we may be able to do it in one go ...

luke-li-2003 commented:
Newest performance numbers, including non-zero second dimensions

|a[n][0]|baseline|accelerated|
|-|-|-|
|0|2.77|257|
|1|1.87|137|
|2|1.47|110|
|16|0.36|21.6|

|a[n][1]|baseline|accelerated|
|-|-|-|
|1|1.9|128|
|2|1.5|102|
|16|0.36|21.1|

|a[n][16]|baseline|accelerated|
|-|-|-|
|1|1.9|85|
|2|1.5|55|
|16|0.34|7.7|

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,592 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+   TR::Register *dimsPtrReg = cg->evaluate(node->getFirstChild());
+   // number of dimensions - compile time constant
+   uint32_t nDims = node->getSecondChild()->get32bitIntegralValue();
+   // class pointer of objects in the array - in the 2D case this is the class of the 1D array
```
reference to 1st D or 2nd D is ambiguous. maybe top level vs. bottom level is clearer? or, clarify it with 1st D (or 1D) referring to the top level?

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,592 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+   // oolJumpLabel is a common point that all branches will jump to.
+   // From this label, we branch to OOL code.
+   // We do this instead of jumping directly to OOL code from mainline
+   // since according to the x and z implementation that is not possible
```
what is the reason it isn't possible?  any insights here?

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,592 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+   endLabel->setEndInternalControlFlow();
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   // if the second dimension's length is not zero, we call the function to handle it
```
It should implemented below, this comment is outdated.

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
This logic flow comes from the original x implementation and it is indeed convoluted:

1. the alignment compensation is for the reference fields (the data), not the header; the header's size is always aligned.

2. since we don't know the length on compile time, we need to figure out the alignment padding on runtime.

E.g. say we want the data to be 8-byte aligned, but referenceFieldSize is 4 bytes; we set alignCompensation to 8-1=7 (line 1965), with -7 being a convenient mask 0b111...1000. Then, calculate `((length * referenceFieldSize) + alignmentCompensation) & (-alignementCompensation)`, in other words, first `addi` (line 2106), then `rlwinm` (line 2107) to get the next aligned size for the data array.

I then noticed we are essentially adding two constants (header size and compensation), so I figured I could save an instruction by just using one addi to add in both, even though they were not related.

zl-wang replied:
but the existing code was written in such a way that there seemed twice alignments on a single array. either useless instruction or unnecessary complication ...

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
you have access to the top-dimension's componentType (i.e. bottom-dimension's array type). from there, you can get its componentType ... then you need to map that type to the proper size (i.e. bottom-dimension's element size): byte/char/short/int/long/float/double/reference;

i understood now why previously it can only deal with 0-length bottom-dimension (because it was wrong).
Look up element size through `arrayElementSize`, or I have impression previously you had hard-coded mapping when you did something else (hashcode?).

luke-li-2003 replied:
In `hashCodeHelper`, the element type was passed in as a child of the call node. I never worked with `arrayElementSize` before, will look it up.

luke-li-2003 replied:
Also, just to double check, the arrayElementSize should be known at compile time in all scenario right? I can't imagine a case where it will not be known.

luke-li-2003 commented:
Well, I know the node (and the register) that will eventually be evaluated into the arrayClass, but every piece of code I could find the resolve the arrayClass itself on compile time ended up being quite long.

zl-wang commented:
> Well, I know the node (and the register) that will eventually be evaluated into the arrayClass, but every piece of code I could find the resolve the arrayClass itself on compile time ended up being quite long.

good point.  the 2D array class could be unresolved under certain circumstances. i am fine then with your chasing the signature for the element size.

luke-li-2003 commented:
The code is working pretty well now, any other comments I haven't addressed?

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,624 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+   TR::Compilation *comp = cg->comp();
+   TR_ASSERT_FATAL(comp->target().is64Bit(),
+      "immMultiplicatoinWithShift is only supported on 64-bit JVMs!");
+
```
you are assuming sourceReg is an unsigned 32bit value, otherwise this isn't entirely correct for every scenario.

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,624 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, targetReg, sourceReg, sourceReg);
+         break;
+      case 1:
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, targetReg, sourceReg);
```
Should I instead create two separate branches in the caller of this helper, where the outcome is expected in `targetReg` for cases bigger than 1, and in `sourceReg` if the case is 1? It's possible, but complicated.

zl-wang replied:
i am wondering if you need this routine at all. for `imm!=1`, i think you don't need a call but just gen a single instruction with derived shiftAmount and corresponding mask:  trailingZeroes() existing already
```
       shiftAmount = trailingZeroes(imm);
       mask = ((1 << (32+shiftAmount)) - 1) & ((-1) << shiftAmount);   // maybe can simplify this further
```

zl-wang replied:
you can certainly look up the existing code how we did this same thing (i think we must have done this before).

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,624 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+
+   // the maximum number of elements we can handle without risking an overflow
+   uintptr_t maxObjectSizeInElements = cg->getMaxObjectSizeGuaranteedNotToOverflow() / referenceFieldSize;
+
```
as i commented before, the logic below cannot be followed easily.  why should we deal with objectAlignment requirement on referenceFieldSize? our ultimate goal is to calculate the (individual) final object (be it java array or normal object) sizes in such a way that each size is multiple of `TR::Compiler->om.getObjectAlignmentInBytes()`  (either 8 or 16 currently).

for each object,   `((headerSize + dataSize) + (alignmentInBytes - 1)) & (-alignmentInBytes)`, that is what in my mind, instead of looking at referenceFieldSize alone.  Basically, I don't understand how looking at fieldSize alone is related to fulfilling the final size alignment.

zl-wang replied:
that is obviously a wrong assumption and it can happen in major ways. example: gencon header is two slots (i.e. compressedRef mode, two slots are 8-byte). but, when heap size is bigger than 32GB, compression/decompression needs to shift by 4 (i.e. requiring 16-aligned). now, the header is not multiple of alignment anymore.

zl-wang replied:
looks like x/z existing implementations have multiple bugs:  leafElementSize mapping is one (except you only deal with 0-length array), and this header size assumption is another.

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
updated

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,611 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+         || (nonZeroArraySizeAligned != TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+   // if a non-zero array needs alignment at runtime, we first add alignmentCompensation,
+   // then mask with (-alignmentInBytes)
+   int32_t nzArrayAlignmentCompensation = needsAlignNonZeroArray ? alignmentInBytes - 1 : 0;
```
as i see it, you need 3 compensation booleans for:  zero-length array, top-dimension nz array, and nz leafArray. 
but you only need 1 compensation amount, i.e. as long as a certain array needs alignment, use the same compensation amount:  (alignmentInBytes - 1).

luke-li-2003 replied:
There is no need to compensate for any zero-length array at runtime, since its length is just the length of the header.

There shouldn't be a scenario where a top-dimension non-zero array needs compensation whilst a leaf array does not, nor vice-versa. 

Which is why as you can see in my code, I have a single boolean and a single compensation amount in the end. 

I tried to make every single expression somewhat readable, but the downside was that I got 4 variables that were only used once, perhaps instead of:

```
   int32_t refFieldSizeAligned = OMR::align(referenceFieldSize, alignmentInBytes);
   bool needsAlignRefField = (refFieldSizeAligned != referenceFieldSize);
   int32_t contHeaderSizeAligned = OMR::align(TR::Compiler->om.contiguousArrayHeaderSizeInBytes(),
                                              alignmentInBytes);
   bool needsAlignNonZeroArray = needsAlignRefField
         || (contHeaderSizeAligned != TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
   int32_t nzArrayAlignmentCompensation = needsAlignNonZeroArray ? alignmentInBytes - 1 : 0;
```

I can do:

```
int32_t nzArrayAlignmentCompensation =
      (OMR::align(referenceFieldSize, alignmentInBytes) != referenceFieldSize)
            || (OMR::align(TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), alignmentInBytes))
                  != TR::Compiler->om.contiguousArrayHeaderSizeInBytes())
      ? alignmentInBytes-1 : 0
```

I decided the former is easier to follow, but I can certainly change to the latter if that's more in-keeping with our coding style.

zl-wang replied:
not right.
you always need to runtime-deciding the zero-length array size (it could be 16-byte/24-byte/32-byte, depending on compression-or-not and heap size).

> There shouldn't be a scenario where a top-dimension non-zero array needs compensation whilst a leaf array does not, nor vice-versa.

this notion is completely wrong to me.  top-dimension element is java-reference, while leaf-array element can be anything (say, byte).  how can the two have definitely same alignment requirement?

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,611 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+   int32_t contHeaderSizeAligned = OMR::align(TR::Compiler->om.contiguousArrayHeaderSizeInBytes(),
+                                              alignmentInBytes);
+   bool needsAlignNonZeroArray = needsAlignRefField
+         || (nonZeroArraySizeAligned != TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
```
typo fixed

ehsankianifar commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
I believe you can use the [`flattenedElementSize` property of the array class](https://github.com/eclipse-openj9/openj9/blob/1f597d25fe99cd920a10dd9bf9eda038dbc9e923/runtime/oti/j9nonbuilder.h#L3644) to get the component size. 
For the first dim array it is safe to use the reference field size (as you already did) but for the leaf element size we have the element size in `flattenedElementSize`. It the node is static you can get the size in compile time like this:
```
J9ArrayClass* clazz = (J9ArrayClass*)node->getThirdChild()->getSymbol()->getStaticSymbol()->getStaticAddress();
J9ArrayClass* compClazz = (J9ArrayClass*)clazz->componentType;
uint32_t componentSize = (uint32_t)compClazz->flattenedElementSize;
```
@zl-wang can you confirm that it is a correct approach?
[I tested it here](https://github.com/ehsankianifar/openj9/blob/dfbfe21e938cd3d719c8f217e5f8bf224df78bad/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L4811-L4813) and it works! (This is a test branch and the link may not work if I delete the branch)

zl-wang replied:
@ehsankianifar Luke and I discussed that before. it doesn't work in case the array class isn't resolved yet. it is very rare for sure, but it indeed can happen i imagined.

luke-li-2003 replied:
@ehsankianifar in your sample code, I have some doubts over directly using the symref from the child node (rather than iterating over the load chain). Taking the componentSize after getting the symref seems fair, though I am not sure about the pros and cons compared to getting the signature out of the symref.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
for Power codegen, vmThreadReg is a locked register, i.e.  it doesn't participate in register assignment.  Just like the java stack pointer register, you don't need to put it on dependencyConditions.

Also, wondering if you really need these many different registers.  could some of them have distinct live-ranges and be re-used in the between ... thus reducing the number of registers being used?  have you checked this out?

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
you probably need heapAlloc and heapTop values in at least 2 places.  do you load them up separately in these 2 places? for better performance, you can load them up, no matter if firstDimLength is zero or not. Just a thought, please consider it.

luke-li-2003 replied:
They are loaded once in each of the three branches ([0][0], [n][0], [n][n]), but performance-wise they should be loaded only once since the code can only reach one branch.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
from reading how you jump around, i guessed topDimension array size calculation, which could be re-used, is not re-used. you will do the calculation all over again? look into how you can improve that.  Did z implementation do this re-calculation as well?  @ehsankianifar

luke-li-2003 replied:
Noted, I made the update.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,582 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+   TR::Register *dimsPtrReg = cg->evaluate(node->getFirstChild());
+   // number of dimensions - compile time constant
+   uint32_t nDims = node->getSecondChild()->get32bitIntegralValue();
+   // class pointer of objects in the array - in the 2D case this is the class of the subarray
```
this comment looks not right:  you directly store classReg to firstDim array's clazz field.  That means classReg is firstDim's array class, instead of subarray's class.

luke-li-2003 replied:
I think we are on the same page here, it's the class of the subarray, not the class of the component of the subarray.

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
I removed them now.

luke-li-2003 commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
Again, I'm not sure about it other than to say it's copied from the x/z implementation.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -1930,13 +1930,555 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+
+   // the maximum number of reference fields in the outer array
+   uintptr_t maxObjectSizeInRefFields = cg->getMaxObjectSizeGuaranteedNotToOverflow() / referenceFieldSize;
+   // the maximum number of elements in the inner array
```
I don't want you to waste instructions on these kind of useless big constants.  use the single constant of `maxObjectSizeInRefFields` as the guard is good enough to me.  any reasonable constant will do actually. re-materializing different constants is a total waste. imagine a max TLH of 128KB or even 2MB, this guard testing never catch the overflow ... the overflow will be caught at heapTop comparison.

luke-li-2003 replied:
Updated

luke-li-2003 commented:
Performance on dacapo graphchi using the same long option [from z](https://github.ibm.com/runtimes/openj9-jit-z/issues/1010#issuecomment-129119362): `java -Djava.security.manager=allow -Xmx2g -Xms2g -Xmn1g -Xshareclasses:none "-Xjit:perfTool,disableEscapeAnalysis,disableLoopTransfer,disableLoopReplicator,disableLoopStrider,acceptHugeMethods" -Xtune:throughput -jar dacapo-23.11-MR2-chopin.jar -n 40 graphchi`

Numbers are benchmark time in msecs; lower is better

Forty iterations:

<img width="1726" height="996" alt="image" src="https://github.com/user-attachments/assets/1e33575b-52dd-4bc4-b8f8-2fd96ba81504" />

Averages:

|Build|time|percentage|
|-|-|-|
|baseline|10767|100%|
|inliner|9748|90.5%|

luke-li-2003 commented:
The updated zero-second-dimension case had no discernable impact on throughput, but at least nothing broke.

luke-li-2003 commented:
The last build will crash, should be fixed now

zl-wang reviewed (approved):
LGTM

luke-li-2003 commented:
The code is working, finally.

The problem was that the size of an array header is not the power of 2 as I assumed when using offheap, so I changed the shift instruction into a multiply instruction.

I also updated the final register to duplicate what was done on x, since before then that instruction would just be optimised out.

zl-wang commented:
Jenkins test sanity aix,plinux jdk8,jdk21
